File: chapter-2/go/insertion_sort.go
package insertion_sort

import "golang.org/x/exp/constraints"

type Less[T any] func(t1, t2 T) bool

func InsertionSort[T any](array []T, less Less[T]) {
	for i := 1; i < len(array); i++ {
		key := array[i]
		j := i - 1

		for j >= 0 && less(key, array[j]) {
			array[j+1] = array[j]
			j--
		}

		array[j+1] = key
	}
}

func Min[T constraints.Integer](t1, t2 T) bool {
	return t1 < t2
}

File: chapter-2/insertion-sort/python/insertion_sort.py
def insertion_sort(array):
    for i in range(1, len(array)):
        key = array[i]
        j = i

        while j > 0 and array[j - 1] > key:
            array[j] = array[j - 1]
            j -= 1

        array[j] = key
